Added doc on testing a pull request. (#2307)
Islandora-CLAW/islandora: docs/contributing/testing-a-pull-request.md

### Step 1: Add a repository (if necessary)

If your code is on a fork, then you will need to add a repository to 
Composer so that it knows where to get your package.

If the code that you want to test is a different branch/tag on the same 
repository that you're currently getting your code from, then you do not 
need to add a repository.

Prepare the following replacement tokens before adding a repository:

* `REPO_NAME`: a name for this repository (mandatory if using the composer 
  one-liner), e.g. `rosiel-islandora`
* `REPO_URL`: the URL to the repository, e.g. `https://github.com/rosiel/islandora`

```shell
# Add custom repo
composer config repo.REPO_NAME vcs REPO_URL
```
Your `composer.json` file should now contain

```json
{
  ...
  "repositories": {
    ...
    "REPO_NAME": {
      "type": "vcs",
      "url": "REPO_URL"
    }
  }
}
```

!!! note "Order of precedence of repositories"
    If you have a matching version spec (e.g. branch name) that's available 
    from multiple repositories, e.g. both islandora's Gitlab and your personal 
    fork both have an `enable-hocr` branch, then the repository that's first 
    in the list in composer.json will take precedence.

> ### Step 2: Require theß custom branch

Reviewer: ```suggestion
### Step 2: Require the custom branch
```

# Testing a Pull Request (i.e. Running Custom Code)

If you are testing a pull request, or for other reasons need to replace the
"official" code with code that's in a PR, or a different branch or fork, this
page offers three methods: using Composer Patches, using Composer to
require the branch and/or fork or installing source repositories with Composer.

This documentation applies to Drupal modules, themes, and recipes, or any
other project that is managed by Composer.

!!! note "Can't I just `git clone`?"
    When managing a non-developmental site with Composer, it's very fragile to
    use commands like `git clone` to obtain code, as it can be overwritten by
    a `composer update`.

    For those who are comfortable cloning code onto a development environment, refer to

>     [Installing source repositories with Composer](#installing-source-repositories-with-composer).

Reviewer: ```suggestion
    [Installing Git repositories with Composer](#installing-git-repositories-with-composer).
```

## Applying a Patch using Composer Patches

This method is best for testing pull requests, because it's very easy to get a
patch from a pull request. If the desired code is not the subject of a
PR, you can still use this method but must generate a `.patch` file yourself.
You may wish to use the other method, as it is more dynamic - see

> section **Using Composer to require a fork or branch**.

Reviewer: ```suggestion
You may wish to use the [fork or branch](#using-composer-to-require-a-fork-or-branch)
method, as it is more dynamic.
```

`drupal/islandora`
* `MY_BRANCH_NAME`: the name of the branch you want to run. Example:
  `testing-fedora-6`. Note that in the case that your branch name is
  "version-like" for example `2.x`, then the `dev` goes at the end, as in
  `2.x-dev`, instead of preceding the branch name as in the template above.

However, if your component is a dependency of another component, then you
will probably need to use an alias. This allows your custom code to "act as"
a version that will meet the requirements of your other component. For
example, if the `drupal/islandora_mirador` package requires
`drupal/islandora:^2.4.1`, then using Composer to require the `enable-hocr`
branch of `drupal/islandora` will not meet the requirements. Instead, use
`as` to provide an alias, to a version that will match the constraints. Note
the quotes around the package name and version spec. This takes the form:

```shell
composer require "MY_PACKAGE:dev-MY_BRANCH_NAME as ALIAS"
```
For example:
```shell
composer require "drupal/islandora:dev-enable-hocr as 2.12.1"
```

That will install the specified branch and allow it to work with your
dependencies.

> ## Installing git repositories with Composer

Reviewer: ```suggestion
## Installing Git repositories with Composer
```

> This method pulls the source (i.e. git) repositories directly into your project

Reviewer: ```suggestion
This method pulls the source (i.e. Git) repositories directly into your project
```

and will allow you to pull in open pull requests by simply following a normal
Git workflow.

!!!note
    If the site has already been installed without using `--prefer-source` you
    will need to clear Composer's cache via `composer clearcache` or including
    `--no-cache` for any `install` or `reinstall` commands (as below).

### Step 1: Re-install the code from source (if required)
Assuming that the environment has not been installed with `--prefer-source`,
reinstall the package.

```shell
composer reinstall MY_PACKAGE --prefer-source
```
with the following replacements:

* `MY_PACKAGE`: the full Composer name of the package. Example:
  `drupal/islandora`

This will pull the code from the source repository (i.e. including the `.

> git` files) and add the package code at the same version of the `MY_PACKAGE`

Reviewer: ```suggestion
This will pull the code from the source repository (i.e. including the `.git`
files) and add the package code at the same version of the `MY_PACKAGE`
```
